### PR TITLE
JAMES-3516 Implement Thread Cassandra table

### PR DESCRIPTION
**Pull request by quantranhong1999**
WIP: tests to be written

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraThreadDAO.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    public Flux<Void> insertSome(Username username, Set<MimeMessageId> mimeMessageIds, MessageId messageId, ThreadId threadId, Optional<Subject> baseSubject) {
+        return Flux.fromIterable(mimeMessageIds)
+            .flatMap(mimeMessageId -> executor.executeVoid(insertOne.bind()
+                .setString(USERNAME, username.asString())
+                .setString(MIME_MESSAGE_ID, mimeMessageId.getValue())
+                .setUUID(MESSAGE_ID, ((CassandraMessageId) messageId).get())
+                .setUUID(THREAD_ID, ((CassandraMessageId) threadId.getBaseMessageId()).get())
+                .setString(BASE_SUBJECT, baseSubject.map(Subject::getValue).orElse(null))));
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraThreadDAO.java:88]: Specify concurrency upon flatmap

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraThreadDAO.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    public Flux<Pair<Optional<Subject>, ThreadId>> selectSome(Username username, Set<MimeMessageId> mimeMessageIds) {
+        return Flux.fromIterable(mimeMessageIds)
+            .flatMap(mimeMessageId -> executor
+                .executeSingleRow(selectOne.bind()
+                    .setString(USERNAME, username.asString())
+                    .setString(MIME_MESSAGE_ID, mimeMessageId.getValue()))
+                .map(this::readRow));
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraThreadDAO.java:97]: Idem

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraThreadDAO.java (hunk 1 above)
Comment (quantranhong1999) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraThreadDAO.java:88]: Do you mean I should put CassandraAsyncExecutor.execute... beyond?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraThreadDAO.java [hunk 3]
```diff
@@ -85,7 +86,7 @@ public CassandraThreadDAO(Session session) {
                 .setString(MIME_MESSAGE_ID, mimeMessageId.getValue())
                 .setUUID(MESSAGE_ID, ((CassandraMessageId) messageId).get())
                 .setUUID(THREAD_ID, ((CassandraMessageId) threadId.getBaseMessageId()).get())
-                .setString(BASE_SUBJECT, baseSubject.map(Subject::getValue).orElse(null))));
+                .setString(BASE_SUBJECT, baseSubject.map(Subject::getValue).orElse(null))), ReactorUtils.DEFAULT_CONCURRENCY);
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraThreadDAO.java:89]: Static import

File: server/container/guice/cassandra/src/main/java/org/apache/james/modules/mailbox/CassandraMailboxModule.java [hunk 4]
```diff
@@ -223,6 +226,7 @@ protected void configure() {
         cassandraDataDefinitions.addBinding().toInstance(CassandraModSeqModule.MODULE);
         cassandraDataDefinitions.addBinding().toInstance(CassandraSubscriptionModule.MODULE);
         cassandraDataDefinitions.addBinding().toInstance(CassandraUidModule.MODULE);
+        cassandraDataDefinitions.addBinding().toInstance(CassandraThreadModule.MODULE);
```
Comment (chibenwa) [inline server/container/guice/cassandra/src/main/java/org/apache/james/modules/mailbox/CassandraMailboxModule.java:229]: Can we have a distinct `CassandraThreadIdGuessingModule` ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraThreadDAO.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                .executeSingleRow(selectOne.bind()
+                    .setString(USERNAME, username.asString())
+                    .setString(MIME_MESSAGE_ID, mimeMessageId.getValue()))
+                .map(this::readRow), ReactorUtils.DEFAULT_CONCURRENCY)
+            .distinct();
+    }
+
+    public Flux<Void> deleteSome(Username username, Set<MimeMessageId> mimeMessageIds) {
+        return Flux.fromIterable(mimeMessageIds)
+            .flatMap(mimeMessageId -> executor.executeVoid(deleteOne.bind()
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraThreadDAO.java:104]: And this one?

File: server/container/guice/cassandra/src/main/java/org/apache/james/modules/mailbox/CassandraMailboxModule.java [hunk 6]
```diff
@@ -226,7 +226,7 @@ protected void configure() {
         cassandraDataDefinitions.addBinding().toInstance(CassandraModSeqModule.MODULE);
         cassandraDataDefinitions.addBinding().toInstance(CassandraSubscriptionModule.MODULE);
         cassandraDataDefinitions.addBinding().toInstance(CassandraUidModule.MODULE);
-        cassandraDataDefinitions.addBinding().toInstance(CassandraThreadModule.MODULE);
+        cassandraDataDefinitions.addBinding().toInstance(CassandraThreadIdGuessingModule.MODULE);
```
Comment (chibenwa) [inline server/container/guice/cassandra/src/main/java/org/apache/james/modules/mailbox/CassandraMailboxModule.java:229]: I meanm can we have a dedicated Guice module <3 ?

Comment (quantranhong1999) [reply server/container/guice/cassandra/src/main/java/org/apache/james/modules/mailbox/CassandraMailboxModule.java:229]: do you mean separate it like this?
image

Comment (quantranhong1999) [reply server/container/guice/cassandra/src/main/java/org/apache/james/modules/mailbox/CassandraMailboxModule.java]: Or I should put ThreadId module here?
image

Comment (chibenwa) [reply server/container/guice/cassandra/src/main/java/org/apache/james/modules/mailbox/CassandraMailboxModule.java]: Yes. To be honnest I think in some configuration supporting threads does not make sense.

(IMAP only)

Having it separated makes it easier to switch it off if needed.

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Can you squask, @quantranhong1999 ?

**Comment by quantranhong1999** [comment]
Just a rebase on master.

**Comment by quantranhong1999** [comment]
```
org.apache.james.mailbox.cassandra.quota.CassandraPerUserMaxQuotaManagerTest.org.apache.james.mailbox.cassandra.quota.CassandraPerUserMaxQuotaManagerTest
java.lang.ExceptionInInitializerError
Caused by: org.testcontainers.containers.ContainerLaunchException: Container startup failed
Caused by: org.testcontainers.containers.ContainerFetchException: Can't get Docker image: RemoteDockerImage(imageName=<resolving>, imagePullPolicy=DefaultPullPolicy())
Caused by: com.github.dockerjava.api.exception.DockerClientException: Could not build image: unable to find image "sha256:57ef4acff3e9ea0fcfe1b36ff7ead08afd9e1b5292353d3e875b1472f6f85028"
```
This seems not related.

**Pull request merged**